support/Python/Machines.py

```diff
@@ -0,0 +1,83 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+"""Support for host machines, such as supercomputers.
+
+Machines are defined as YAML files in 'support/Machines/'. To add support for a
+new machine, add a YAML file that defines a `Machine:` key with the attributes
+listed in the `Machine` class below.
+
+To select a machine, specify the `MACHINE` option when configuring the CMake
+build.
+"""
+
+import os
+from dataclasses import dataclass
+from typing import Optional
+
+import yaml
+
+
+@dataclass(frozen=True)
+class Machine(yaml.YAMLObject):
+    """A machine we know how to run on, such as a particular supercomputer.
+
+    Many configuration options for job submission are hardcoded in the submit
+    script for the machine (such as the total number of available cores per
+    node). Here we provide parameters that we want to adjust when scheduling
+    jobs, but that have sensible defaults based on our experience running on the
+    machine. Parameters are set to 'None' when they have no sensible default for
+    the machine.
+
+    Attributes:
+      Name: A short name for the machine. Must match the YAML file name.
+      Description: A description of the machine. Give some basic context and
+        any information that may help people get started using the machine.
+        Provide links to wiki pages, signup pages, etc., for additional
+        information.
+      DefaultProcsPerNode: Default number of worker threads spawned per node.
+        It is often advised to leave one core per node or socket free for
+        communication, so this might be the number of cores or hyperthreads
+        per node minus one.
+      DefaultQueue: Default queue that jobs are submitted to. On Slurm systems
+        you can see the available queues with `sinfo`.
+      DefaultTimeLimit: Default wall time limit for submitted jobs. For
+        acceptable formats, see: https://slurm.schedmd.com/sbatch.html#OPT_time
+    """
+    yaml_tag = '!Machine'
+    yaml_loader = yaml.SafeLoader
+    # The YAML machine files can have these attributes:
+    Name: str
+    Description: str
+    DefaultProcsPerNode: Optional[int] = None
+    DefaultQueue: Optional[str] = None
+    DefaultTimeLimit: Optional[str] = None
```

Comment:
[Can't continue the previous thread, so making a new one]

Personally, I think these are all things we should have to specify in the `MACHINE.yaml` file (as defaults that could potentially be overridden). If you want a default-created submit script for a machine (like you mentioned above, not having to input number of worker threads every time you schedule a job), you'll need actual values for these, not `None`.

Reply:
I'm not sure I understand you correctly. These values can currently be specified in the machine file, and then they'll be used as default. But they don't have to be in case there's no obvious default (like maybe the machine has multiple queues and you always want the user to choose or something).

Reply:
Ah, ok. I *do* think there are some obvious defaults for these for each machine :) (even the ones with multiple queues). But I don't think this is worth holding up this PR. Leave it as is for now.

Reply:
You might be right that we'll always have sensible defaults for these. I removed the optional for now and added some tests. If we encounter a machine where we don't have a sensible default, we can make the field optional.